### PR TITLE
Creating a data profiler to profile data frames.

### PR DESCRIPTION
**Pull request by sunishsheth2009**
## What changes are proposed in this pull request?
Creating a data profiler to profile data frames.

## How is this patch tested?
- [x] Running 
```
from mlflow.pipelines.cards import pandas_renderer
pandas_renderer.render_html(df)
``` 
with different df.

## Does this PR change the documentation?

- [x] No. You can skip the rest of this section.
- [ ] Yes. Make sure the changed pages / sections render correctly by following the steps below.

1. Click the `Details` link on the `Preview docs` check.
2. Find the changed pages / sections and make sure they render correctly.

### Is this a user-facing change?

- [x] No. You can skip the rest of this section.
- [ ] Yes. Give a description of this change to be included in the release notes for MLflow users.

(Details in 1-2 sentences. You can just refer to another PR with a description if this PR is part of a larger change.)

### What component(s), interfaces, languages, and integrations does this PR affect?
Components
- [ ] `area/artifacts`: Artifact stores and artifact logging
- [ ] `area/build`: Build and test infrastructure for MLflow
- [ ] `area/docs`: MLflow documentation pages
- [ ] `area/examples`: Example code
- [ ] `area/model-registry`: Model Registry service, APIs, and the fluent client calls for Model Registry
- [ ] `area/models`: MLmodel format, model serialization/deserialization, flavors
- [x] `area/pipelines`: Pipelines, Pipeline APIs, Pipeline configs, Pipeline Templates
- [ ] `area/projects`: MLproject format, project running backends
- [ ] `area/scoring`: MLflow Model server, model deployment tools, Spark UDFs
- [ ] `area/server-infra`: MLflow Tracking server backend
- [ ] `area/tracking`: Tracking Service, tracking client APIs, autologging

Interface
- [ ] `area/uiux`: Front-end, user experience, plotting, JavaScript, JavaScript dev server
- [ ] `area/docker`: Docker use across MLflow's components, such as MLflow Projects and MLflow Models
- [ ] `area/sqlalchemy`: Use of SQLAlchemy in the Tracking Service or Model Registry
- [ ] `area/windows`: Windows support

Language
- [ ] `language/r`: R APIs and clients
- [ ] `language/java`: Java APIs and clients
- [ ] `language/new`: Proposals for new client languages

Integrations
- [ ] `integrations/azure`: Azure and Azure ML integrations
- [ ] `integrations/sagemaker`: SageMaker integrations
- [ ] `integrations/databricks`: Databricks integrations

<a name="release-note-category"></a>
### How should the PR be classified in the release notes? Choose one:

- [ ] `rn/breaking-change` - The PR will be mentioned in the "Breaking Changes" section
- [x] `rn/none` - No description will be included. The PR will be mentioned only by the PR number in the "Small Bugfixes and Documentation Updates" section
- [ ] `rn/feature` - A new user-facing feature worth mentioning in the release notes
- [ ] `rn/bug-fix` - A user-facing bug fix worth mentioning in the release notes
- [ ] `rn/documentation` - A user-facing documentation change worth mentioning in the release notes

### CODE REVIEW

File: mlflow/pipelines/cards/histogram_generator.py [hunk 1]
```diff
@@ -0,0 +1,124 @@
...
+    be ordered and corresponding to equal distant percentiles.
+
+    :param quantiles: The quantiles that capture the frequency distribution.
+    :param num_buckets: The number of buckets in the generated equal height histogram.
+    :return: An equal height histogram or None if inputs are invalid.
+    """
+    if (
+        (len(quantiles) == 0)
+        or (len(quantiles) < 3)
+        or (len(quantiles) % 2 == 0)
```
Comment (dbczumar) [inline mlflow/pipelines/cards/histogram_generator.py:19]: Why does the length of quantiles need to be even? E.g. is 3 buckets demarcated by 4 quantiles invalid?

File: mlflow/pipelines/cards/histogram_generator.py [hunk 2]
```diff
@@ -0,0 +1,124 @@
...
+    """Generates the equal height histogram from the input quantiles. The quantiles are assumed to
+    be ordered and corresponding to equal distant percentiles.
+
+    :param quantiles: The quantiles that capture the frequency distribution.
+    :param num_buckets: The number of buckets in the generated equal height histogram.
+    :return: An equal height histogram or None if inputs are invalid.
+    """
+    if (
+        (len(quantiles) == 0)
+        or (len(quantiles) < 3)
```
Comment (dbczumar) [inline mlflow/pipelines/cards/histogram_generator.py:18]: Nit: `len(quantiles) == 0` is redundant here

File: mlflow/pipelines/cards/histogram_generator.py (hunk 1 above)
Comment (dbczumar) [reply mlflow/pipelines/cards/histogram_generator.py:19]: (I realize we're pulling this in from Databricks, but I'd still like to understand better if possible :D)

File: tests/pipelines/cards/test_pandas_renderer.py [hunk 3]
```diff
@@ -0,0 +1,30 @@
...
+Test to convert data from tabular format to DatasetFeatureStatisticsList proto.
+"""
+import numpy as np
+
+from facets_overview import feature_statistics_pb2
+from mlflow.pipelines.cards import pandas_renderer
+
+# _NUMERIC_TYPES = ("tinyint", "smallint", "int", "bigint", "float", "double", "decimal(10,0)")
+# _TIME_TYPES = ("timestamp", "date")
+# _STRUCT_TYPES = ("array<bigint>", "map<string,bigint>", "struct<col1:int,col2:int>")
```
Comment (dbczumar) [inline tests/pipelines/cards/test_pandas_renderer.py:11]: Can we remove commented code?

File: mlflow/pipelines/cards/histogram_generator.py (hunk 1 above)
Comment (sunishsheth2009) [reply mlflow/pipelines/cards/histogram_generator.py:19]: Spoke to Shupeng Sun. It seems like we can remove them and ease the restrictions. 

We still want to keep `(len(quantiles) < 3)` since we ideally don't want to just show 1 bucket. So still keeping that condition.

File: tests/pipelines/cards/test_pandas_renderer.py (hunk 3 above)
Comment (sunishsheth2009) [reply tests/pipelines/cards/test_pandas_renderer.py:11]: Yup still working on the test

File: mlflow/protos/feature_statistics.proto [hunk 4]
```diff
@@ -0,0 +1,332 @@
...
+// ==============================================================================
+
+// Adapted from https://github.com/PAIR-code/facets/blob/master/facets_overview/proto/feature_statistics.proto
+// - Ported from proto3 to proto2 by adding "optional" modifiers to relevant fields
+// - unsigned int64 types changed to int64 types
+
+// Definitions for aggregated feature statistics for datasets.
+package featureStatistics;
+
+option java_package = "com.databricks.api.proto.featureStatistics";
```
Comment (dbczumar) [inline mlflow/protos/feature_statistics.proto:25]: Can we change the package name here? Maybe reuse the facets package name?

File: mlflow/pipelines/cards/pandas_renderer.py [hunk 5]
```diff
@@ -0,0 +1,244 @@
...
+            feat_stats = feat.string_stats
+            strs = []
+            compute_unique_str = current_column_value.dropna()
+            for item in compute_unique_str:
+                strs.append(
+                    item
+                    if hasattr(item, "__len__")
+                    else item.encode("utf-8")
+                    if hasattr(item, "encode")
+                    else str(item)
```
Comment (jinzhang21) [inline mlflow/pipelines/cards/pandas_renderer.py:142]: Will these 2 lines ever be called?

File: mlflow/pipelines/cards/pandas_renderer.py [hunk 6]
```diff
@@ -0,0 +1,244 @@
...
+    :param column: A column from a DataFrame.
+    :return: A CommonStatistics proto.
+    """
+    common_stats = facet_feature_statistics_pb2.CommonStatistics()
+    common_stats.num_missing = column.isnull().sum()
+    common_stats.num_non_missing = len(column) - common_stats.num_missing
+    # Default set using: https://src.dev.databricks.com/databricks/universe/-/blob/model-monitoring/python/databricks/model_monitoring/rendering/html_renderer.py?L33-L35&subtree=true
+    common_stats.min_num_values = 1
+    common_stats.max_num_values = 1
+    common_stats.avg_num_values = 1.0
```
[the remark this answers is not in the document]
Comment (sunishsheth2009) [reply mlflow/pipelines/cards/pandas_renderer.py:71]: I went through this code here: 
https://github.com/PAIR-code/facets/blob/4742b8b93c2dacf22fc8ace2cee42dd06382c48e/facets_overview/facets_overview/base_generic_feature_statistics_generator.py#L290-L304

And understanding this code I realized that these are the values is what it was computed to. 

And then I realized that this is also done exactly the same way here: https://src.dev.databricks.com/databricks/universe/-/blob/model-monitoring/python/databricks/model_monitoring/rendering/html_renderer.py?L33-L35&subtree=true

Do you think we should replicate what facet-overview is doing or it is not required? It seems to be working for our databricks profiling as well.

File: mlflow/pipelines/cards/pandas_renderer.py (hunk 5 above)
Comment (sunishsheth2009) [reply mlflow/pipelines/cards/pandas_renderer.py:142]: Got this code from here: https://github.com/PAIR-code/facets/blob/4742b8b93c2dacf22fc8ace2cee42dd06382c48e/facets_overview/facets_overview/base_generic_feature_statistics_generator.py#L263-L265
I assumed that this might be required for an edge case when the strings are not encoded. So left this code path here. 

Do you think we need to support this usecase? I am happy to remove it

File: tests/pipelines/cards/test_pandas_renderer.py [hunk 7]
```diff
@@ -0,0 +1,250 @@
...
+                  high_rank: 2
+                  label: "NFLX"
+                  sample_count: 1.0
+                }
+              }
+            }
+          }
+          features {
+            name: "Shares"
+            type: FLOAT
```
[the remark this answers is not in the document]
Comment (sunishsheth2009) [reply tests/pipelines/cards/test_pandas_renderer.py:131]: Here is what I found:
``` 
0    100
1    170
2    150
3    200
4    123
Name: Shares, dtype: int64
```

And 
```
0    100.0
1    170.0
2    150.0
3    200.0
4      NaN
Name: Shares, dtype: float64
```

So the None value in the column makes it a float.

File: mlflow/pipelines/cards/pandas_renderer.py [hunk 8]
```diff
@@ -0,0 +1,244 @@
...
+
+from typing import Union, Iterable, Tuple
+from mlflow.protos import facet_feature_statistics_pb2
+from mlflow.pipelines.cards import histogram_generator
+
+
+def DtypeToType(dtype):
+    """Converts a Numpy dtype to the FeatureNameStatistics.Type proto enum."""
+    fs_proto = facet_feature_statistics_pb2.FeatureNameStatistics
+    if dtype.char in np.typecodes["AllFloat"]:
```
[the remark this answers is not in the document]
Comment (sunishsheth2009) [reply mlflow/pipelines/cards/pandas_renderer.py:17]: Good catch.
Checked how this is handled by databricks profiling and facet_oveview as well. 
They both throw a TypeError that the `Found type complex, but expected one of: int, long, float`
We could replicate that. What do you think?

File: tests/pipelines/cards/test_pandas_renderer.py [hunk 9]
```diff
@@ -0,0 +1,250 @@
...
+from mlflow.pipelines.cards import pandas_renderer
+
+
+def test_convert_to_html():
+    proto = facet_feature_statistics_pb2.DatasetFeatureStatisticsList()
+    html = pandas_renderer.construct_facets_html(proto)
+    assert len(html) != 0
+
+
+def testDTypeToType():
```
[the remark this answers is not in the document]
Comment (sunishsheth2009) [reply tests/pipelines/cards/test_pandas_renderer.py]: Done

File: mlflow/pipelines/cards/pandas_renderer.py (hunk 6 above)
[the remark this answers is not in the document]
Comment (jinzhang21) [reply mlflow/pipelines/cards/pandas_renderer.py:71]: I think we should match the Facet implementation. Setting them to 1 is not the correct approach.

File: mlflow/pipelines/cards/pandas_renderer.py (hunk 5 above)
Comment (jinzhang21) [reply mlflow/pipelines/cards/pandas_renderer.py:146]: No I'm just questioning the usage of 2 consecutive if-else clauses. I doubt the 2nd one will ever get called. In what circumstances will it happen? It'd be great to demonstrate with an example.

File: mlflow/pipelines/cards/pandas_renderer.py (hunk 8 above)
[the remark this answers is not in the document]
Comment (jinzhang21) [reply mlflow/pipelines/cards/pandas_renderer.py]: SGTM. Stats really is all about real numbers anyway.

File: tests/pipelines/cards/test_pandas_renderer.py (hunk 7 above)
[the remark this answers is not in the document]
Comment (jinzhang21) [reply tests/pipelines/cards/test_pandas_renderer.py:157]: Got it. That's probably fine.

File: mlflow/protos/facet_feature_statistics.proto [hunk 10]
```diff
@@ -0,0 +1,329 @@
+syntax = "proto2";
```
Comment (jinzhang21) [inline mlflow/protos/facet_feature_statistics.proto:1]: Can we add a comment to link explicitly to the GH commit to this file we copy from Facet? It'd be helpful to debug in the future if Facet introduces some backward in compatible changes.

File: mlflow/pipelines/cards/pandas_renderer.py (hunk 6 above)
[the remark this answers is not in the document]
Comment (sunishsheth2009) [reply mlflow/pipelines/cards/pandas_renderer.py:71]: Spoke offline to add a TODO to support multi-dimensional columns later on.

**Review by jinzhang21 — APPROVED** [review]
